apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,375 @@
+
+#### 2. Download free-threaded Python
+
+At the start of my internship, Python 3.14t, the version my mentor suggested I use for the project, was in development. I had trouble installing it, so my mentor introduced me to [pyenv](https://github.com/pyenv/pyenv), a Python version manager for macOS and Linux. I hadn't used `pyenv` before, instead preferring to use `conda`, but it was a elegant solution to the problem of managing and installing multiple versions of Python. To install the development version of 3.14t, all it took was running the command `pyenv install 3.14t-dev`.
+
+**Note:** Sometimes I used the GIL-enabled build of 3.14 instead. When I did that, I had to also make sure the environment variables `PYTHON_CONTEXT_AWARE_WARNINGS` and `PYTHON_THREAD_INHERIT_CONTEXT` were set to true. These ensure warnings and context play nicely with threads. They are set to true on free-threaded builds, and false otherwise. The Python docs describe these variables more in-depth [here](https://docs.python.org/3/whatsnew/3.14.html#free-threaded-mode-improvements).
+
+#### 3. Create NumPy fork and clone it to my machine
+
+While I didn't have a lot of experience with OSS, I have used git and GitHub before.
+
+#### 4. Build NumPy locally using spin
+
+To run the NumPy test suite, I had to build it locally first. NumPy recommends using [spin](https://github.com/scientific-python/spin), detailing the steps [here](https://numpy.org/doc/stable/building/index.html). After installing the various system and Python dependencies (preferably under a virtual environment), all it took was running `spin build` to get everything built.
+
+#### 5. Install pytest-run-parallel
+
+And finally, the last step is to get `pytest-run-parallel`. When it comes to Python, I prefer to use virtual environments, setting them up using `venv`. After activating the environment, `pip install pytest-run-parallel` got me the plugin.
+
+---
+
+## Park Three: Discovery
+
+With everything put together, I could finally get started on testing the NumPy test suite, running `spin test -- --parallel-threads=auto`. "spin test" is how NumPy runs the test suite under `pytest` when built under `spin`. "--parallel-threads=auto" is a command line option from `pytest-run-parallel` which activates the plugin, telling it to run each test under the specified number of threads. The keyword `auto` looks at the number of available CPU cores and determines the number of threads for you (for me, it was 24).
+
+Alright, if everything in the test suite was "thread-safe" (aka can run under multiple threads at the same time), then everything should run fine, with no test failures whatsoever! As you might be able to tell, since my project was to improve the thread safety of the test suite, this was not the case.
```

Comment:
```suggestion
Alright, if everything in the test suite was "thread-safe" (aka can run under multiple threads at the same time), then everything should run fine, with no test failures whatsoever! As you might be able to tell, since my project was to improve the thread safety of the test suite, this was not the case at first.
```